faculty/clients/api.py

```diff
@@ -0,0 +1,635 @@
+# Copyright 2018-2021 Faculty Science Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Manage Faculty APIs.
+"""
+from collections import namedtuple
+from enum import Enum
+
+from marshmallow import fields, post_load
+from marshmallow_enum import EnumField
+
+from faculty.clients.base import BaseClient, BaseSchema
+from faculty.clients.server import _ServerSchema as ServerSchema
+
+Instance = namedtuple("Instance", ["instance", "outdated", "key"])
```

Comment:
Newer code should use `attrs` rather than `namedtuple` for new data classes. See, for example, the job client - note in particular that the longer form imports (`from attr import attrs, attrib`) are preferred.

Reply:
@acroz while I was checking this out, came across the more modern use of `attr` with `define` and `field` rather than `attrs` and `attrib` (that are the "classic" versions and actually used underneath). https://www.attrs.org/en/stable/api.html#api-reference

Would it make sense to use the new define/field way for newer code (and possibly migrate the older code to that form as well? It feels like going with the classic form of attrs/attrib just for consistency sake in other modules is inconsistent how the SDK moved to using this in place of namedtuples in the first place (ie. sensible change is good).

Anyways I'm still looking at the differences a bit, the new terminology seems more self-explanatory so far...